### PR TITLE
Allow non-value function args and returns

### PR DESCRIPTION
**Pull request by ccasin**
This allows functions argument and return types to have any representable layout, instead of just `value`.  It will be needed for `#float`.

The first commit makes the changes in the type system and just accepts all the test changes (so you can see clearly what changed).  The second commit then slightly refactors the tests to make more sense in the new world.  The tests are slightly limited owing the lack of an additional sort than can make it through the lambda translation, but I put in CRs describing some of the tests we'll want then.

This restriction is still on arrows, not functions.  Obviously we want to move it, but that is orthogonal and I think not needed for v2.

### Defaulting
We're creating a lot more sort variables now, but I've only added a tiny bit of additional defaulting (in the void sanity checks, to avoid defaulting to void).  The system already has sort variable defaulting for types that appear in interfaces (`Typemod.remove_mode_and_layout_variables`), so nothing changes in that case.

For other sort variables, the plan is not to have some big defaulting pass, but rather that any time we look at a layout in the translation to lambda we make sure to default it.  If there are any sort variables we never look at and which don't appear in interfaces, it's fine to leave them.  The infrastructure fot this is already in place, and used by the current sanity checking.  As we add unboxed floats, that sanity checking will change from erroring if not value to erroring if void.

### CODE REVIEW

File: ocaml/lambda/translcore.ml [hunk 1]
```diff
@@ -51,8 +51,16 @@ let layout_must_be_value loc layout =
...
+   sanity check for value.  But we still default to value before checking for
+   void, to allow for sort variables arising in situations like
+
+     let foo () = raise Foo; ()
+
+   When this sanity check is removed, consider whether we are still defaulting
+   appropriately.
+*)
 let layout_must_not_be_void loc layout =
+  Layout.default_to_value layout;
```
Comment (goldfirere) [inline ocaml/lambda/translcore.ml:63]: This is a small perf concern, but could be fixed by refactoring `default_to_value`.

File: ocaml/lambda/translmod.ml [hunk 2]
```diff
@@ -45,8 +45,16 @@ exception Error of Location.t * error
...
-   remove the safety check. *)
+   remove the safety check. But we still default to value before checking for
+   void, to allow for sort variables arising in situations like a module
+   that is just:
+
+     exit 1;;
+
+   When this sanity check is removed, consider whether it must be replaced with
+   some defaulting. *)
 let layout_must_not_be_void loc ty layout =
```
Comment (goldfirere) [inline ocaml/lambda/translmod.ml:56]: Not really part of this patch, but this function is duplicated with translcore. Maybe move to Layout?

File: ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference [hunk 3]
```diff
@@ -2,7 +2,7 @@ File "main.ml", line 1, characters 8-11:
 1 | let _ = A.a = B.b
             ^^^
 Error: This expression has type M.a but an expression was expected of type
-         ('a : value)
-       M.a has layout any, which is not a sublayout of value.
+         ('a : '_representable_layout_1)
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference:5]: This is ugly. Can we do better? Or maybe this happens only for missing cmis, so maybe it's OK?

File: ocaml/typing/ctype.ml [hunk 4]
```diff
@@ -3747,13 +3747,16 @@ exception Filter_arrow_failed of filter_arrow_failure
 
 let filter_arrow env t l ~force_tpoly =
   let function_type level =
-    (* CR layouts v2: This is one of two primary places where we are restricting
-       function arguments / returns to have layout value.  This one handles
+    (* CR layouts v3: This is one of two primary places where we are restricting
```
Comment (goldfirere) [inline ocaml/typing/ctype.ml:3750]: As discussed, I'd like to keep this v2. Maybe we'll backslide on this in a few weeks, but I'm not quite ready to yet.

File: ocaml/typing/typetexp.ml [hunk 5]
```diff
@@ -478,21 +483,20 @@ and transl_type_aux env policy mode styp =
           let arg_mode = Alloc_mode.of_const arg_mode in
           let ret_mode = Alloc_mode.of_const ret_mode in
           let arrow_desc = (l, arg_mode, ret_mode) in
-          (* CR layouts v2: For now, we require function arguments and returns
-             to have layout value.  See comment in [Ctype.filter_arrow].  *)
+          (* CR layouts v3: For now, we require function arguments and returns
```
Comment (goldfirere) [inline ocaml/typing/typetexp.ml:486]: v2?

**Review by goldfirere — COMMENTED** [review]
Looks good overall.

File: ocaml/lambda/translcore.ml (hunk 1 above)
Comment (ccasin) [reply ocaml/lambda/translcore.ml:63]: Agreed.

I just pushed a commit that refactors the `repr` stuff in layouts, implementing the observation we made a while ago that if we have a separate version of `repr` for the case where we're defaulting, it needn't allocate.  While I was there, I made a couple other simplifications:

- We only ever default to value.  Now you don't have a choice, which simplifies things a little.  While we might change the default to `void` later, we have no plans for a version that defaults to different things in different spots.
- Once those simplifications were made, there was no difference between `repr` and `get`.  So now we just have `get` and `get_default_layout`.  I shuffled the functions around a little to put the related "elimination" and "defaulting" functions together.

I think the result is a nice overall simplification, but yell at me if you think otherwise.

File: ocaml/lambda/translmod.ml (hunk 2 above)
Comment (ccasin) [reply ocaml/lambda/translmod.ml:56]: #1402 changes `translmod` to do something different here, so it's no longer duplicated.  I agree it would be cleaner to move all the functions, but they are an artifact of the void safety check so I'd rather wait until we eliminate that to consider what's the best thing.

File: ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference (hunk 3 above)
Comment (ccasin) [reply ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference:5]: I was kind of hoping your histories patch would magically make this better when merged.  Is that too optimistic?

File: ocaml/typing/ctype.ml (hunk 4 above)
Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: After another discussion, I think the conclusion here is that the questions of (a) what we allow for function args / returns and (b) whether to put that restriction on arrows or functions are orthogonal.  It may turn out that moving the restriction to functions is simple, but I'm not inclined to spend any time on it before we ship floats. 

On the other hand, I do think it's worth trying to allow `any` returns, because otherwise non-returning functions (e.g., any function which is a wrapper around `raise`) can't have a nice type.  I gave this a quick shot just now and predictably ran into various assertion / safety check failures.  I am out of time but will give it a more serious try tomorrow morning.  If there are real complications we can ship without it and follow up after, being slightly sad.

Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: OK, I couldn't resist at least looking at an example before running out the door.   Here's a simplified example of a very sad pattern that occurs in the stdlib:

```ocaml
let zero_out a =
  let next i =
    if i = 0 then raise Done else (i-1)
  in
  let rec go i =
    a.(i) := 0;
    go (next i)
  in
  try go (Array.length a) with Done -> ()
```

And for the translation of lambda wants a layout for the return of `go`, so it is sad.  I'll think about this in the morning.

Comment (Ekdohibs) [reply ocaml/typing/ctype.ml:3750]: In the lambda representation of layouts, we added a layout `Pbottom` for cases such as the result of a raise, where the bottom layout is a sublayout of everything (while it seems to me the layout any is a superlayout of everything?). Wouldn't adding a layout bottom in the typechecker as well work here?

Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: Yes, in our lattice "top"/`any` means we don't know anything about the layout and, if we had a bottom, it would mean that we know the layout can be anything. Our "any" is poorly named. I agree non-returning functions can have layout bottom, morally. We can't easily add this layout in the typechecker though, because of how our inference works (for example, we rely on incompatible layouts having no meet).

One observation is that if we finish checking a function and its return layout is any, this means we didn't need to know anything about the layout to check the function.  So the layout doesn't really matter, and could be anything.  From a typing perspective, although I hadn't thought of it this way until now, maybe it would be better to say that the meaning of a fully generalized `'a` of layout `any` is that it could be anything, while the meaning of `any` during inference is that we don't know anything.

After typechecking, any remaining `any`s represent places where no value will really be produced. So maybe we can translate `any` to any layout supported by lambda.  Or maybe we _have_ to pick bottom: otherwise something funny is happening in the `try ... with` - the layouts of the scrutinee and the branch should need to be compatible.  If we, for example, said the return of `go` has layout `void`, they wouldn't be.

Comment (goldfirere) [reply ocaml/typing/ctype.ml:3750]: For what it's worth, I agree with @lpw25 here. `Pbottom` sounds like the correct translation target here.

File: ocaml/typing/layouts.ml [hunk 6]
```diff
@@ -48,24 +48,36 @@ module Sort = struct
...
-            let t = of_const const in
-            r := Some t;
-            t
-          end
-      end
+      | None -> t
       | Some s -> begin
-          let result = repr ~default s in
+          let result = repr s in
           r := Some result; (* path compression *)
```
Comment (goldfirere) [inline ocaml/typing/layouts.ml:57]: While we're in town: this `Some result` will allocate a new `Some` node, even when there's no path to compress. Might be better to have `if result != s then r := Some result`, but that introduces another branch. I think probably allocation is always worse than a straightforward branch, though.

File: ocaml/typing/ctype.ml (hunk 4 above)
Comment (mshinwell) [reply ocaml/typing/ctype.ml:3750]: Hmm, I'm not sure the middle/backend is really treating the layouts more like types rather than kinds.  To my mind in these IRs the situation is a slight refinement of what @lpw25 said above: for `go` it's not that it is impossible to call - that's fine, but it _is_ impossible to return from it.  You cannot construct a term that reads the return value because the bottom kind is uninhabited.  Indeed, the one place where `Pbottom` is used in the middle end is in `Cmmgen`, and the semantics is to delete everything following.

The Flambda 2 code does not look at `Pbottom`, so once we're fully switched over it can be removed.

File: ocaml/lambda/translcore.ml (hunk 1 above)
Comment (goldfirere) [reply ocaml/lambda/translcore.ml:63]: Happy here. Thanks for doing this.

File: ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference (hunk 3 above)
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-missing-cmi/test.compilers.reference:5]: (Just resolve after you read this.)

File: ocaml/typing/ctype.ml (hunk 4 above)
Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: I don't think Leo meant impossible to call in the sense you couldn't run the code, but rather impossible to call in just because there's no type you can instantiate `'a` with.

Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: On the topic of removing `Pbottom`, I'm not so sure.  We need to pick some (lambda) layout for the return of `go`.  Picking an arbitrary other layout seems odd - if we pick `#float`, for example, we have a `try ... with` where the body had layout `#float` and the branch has layout `immediate`.  It's tempting to pick `void`, but I think then we're still in that situation (oddly incompatible layouts in an expression).

File: ocaml/typing/layouts.ml (hunk 6 above)
Comment (ccasin) [reply ocaml/typing/layouts.ml:57]: Yeah.  I preferred the simpler version because (a) it's not on the path we added calls to and are trying to optimize here (this is the non-defaulting repr, but we added calls to the defaulting one), and (b) the branch may not be well predicted.  But I guess the allocation is potentially long-lived, so probably the other way is better?  I'll change it.

File: ocaml/typing/layouts.ml [hunk 7]
```diff
@@ -48,24 +48,36 @@ module Sort = struct
...
           result
         end
     end
 
+  (* To avoid allocating *)
+  let default_value : t option = Some (Const Value)
```
[the remark this answers is not in the document]
Comment (ccasin) [reply ocaml/typing/layouts.ml:63]: Yes, thanks (this was more complicated in an intermediate version and I neglected to update the comment).  I'll remove it.

File: ocaml/typing/ctype.ml (hunk 4 above)
Comment (Ekdohibs) [reply ocaml/typing/ctype.ml:3750]: It seems to me that since we have a concept of layout subkinding, it would be correct to instantiate `val go ('a : bot). int -> 'a` with any `'a`, since it is only used in covariant positions. Alternatively, while this is forbidden since we have no way to compile it further, `val ignore ('a : any): 'a -> unit` (as a function) could be instantiated with any `'a`, since it is only used in contravariant positions.
Concerning the removal of `Pbottom`, it is needed to be able to construct the layout of an expression, but I think this is again only used in closure and flambda1.

Comment (lpw25) [reply ocaml/typing/ctype.ml:3750]: > it would be correct to instantiate val go ('a : bot). int -> 'a with any 'a

In general, you can do that trick by instantiating `'a` with an empty type and then saying that the empty type is a subtype of all types of all kinds. Here it's not clear if you can since the empty type is not of kind `bot`, so you can't instantiate it with the empty type in the first place.

> val ignore ('a : any): 'a -> unit (as a function) could be instantiated with any 'a

With that type, `ignore` can already be instantiated with any type since all types have kind `any`. The variance of the type variable is not relevant.

Comment (ccasin) [reply ocaml/typing/ctype.ml:3750]: In the unboxed type planning meeting just now, we decided to put off allowing `any` returns for a different PR (Mark observed it will require minor flambda2 changes, and anyway it's probably better to do after the PR that properly hooks type system layouts up to lambda layouts).  We still hope to do it for v2.

**Comment by ccasin** [comment]
@goldfirere I think all comments have now been addressed and this is ready for approval.

**Review by goldfirere — APPROVED** [review]
LGTM.
